### ISSUE TITLE
cls inference

### ISSUE THREAD

**Issue by liusse** [ultralytics/ultralytics#22218]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

The model results after training are shown in Figure 1, and it seems there is no overfitting. However, the performance during inference is not very good. I am not sure if this is related to my classification dataset. My main objective is to distinguish between abnormal images and normal images. The dataset includes the green_bar class (as shown in Figure 2) with about 1,500 images, the green_screen class (as shown in Figure 3) with about 1,500 images, and the nomal class with about 10,000 images. The nomal class and the green_bar class are very easy to misclassify because they are quite similar, making them easy to confuse. Is the imbalance in my dataset ratio causing these issues? This is very frustrating for us. Thank you very much for your help in resolving this.

<img width="642" height="643" alt="Image" src="https://github.com/user-attachments/assets/bc16b2e5-e03b-4357-9d79-e6a64a984e56" />

<img width="710" height="668" alt="Image" src="https://github.com/user-attachments/assets/f6f5b7d6-0a04-4214-9072-71b46237bf43" />

<img width="328" height="336" alt="Image" src="https://github.com/user-attachments/assets/7ac32f47-d94f-4321-add0-43ba7ac77dc7" />

### Additional

The normal class consists of some captured materials, screenshots, and other images that do not contain any anomalies.
I also found that the model is insensitive to color; for example, if an image is entirely purple or blue, it may still be classified as green_screen. I only want an image that is mostly green

**Comment by UltralyticsAssistant** [comment]
👋 Hello @liusse, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started; an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

To help us assist with your classification results and class imbalance concerns, please share:
-Your exact training command or Python snippet, model used (e.g., YOLO11n-cls/YOLO11s-cls), epochs, image size, batch size, and any augmentations
-A brief description of your dataset split (train/val/test counts per class) and a few sample images from each class
-Validation metrics and training logs (accuracy, F1, per-class precision/recall) plus the confusion matrix
-The exact inference command or Python code, confidence threshold, and example inputs/outputs where misclassifications occur
-Confirmation of the Ultralytics version and PyTorch/CUDA environment details

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Accuracy isn't useful when your dataset is imbalanced. You should have at least a balanced validation set.

**Comment by Y-T-G** [comment]
Also are you inferencing using Ultralytics?

**Comment by liusse** [comment]
> Also are you inferencing using Ultralytics?

yes, used classification functionality

**Comment by liusse** [comment]
> Accuracy isn't useful when your dataset is imbalanced. You should have at least a balanced validation set.

okay, I will expand the dataset later
but I have a question: if I use a portion of the images from the normal class to generate abnormal images and then place them into the abnormal class, will this compromise the training effectiveness?

**Comment by glenn-jocher** [comment]
Great question—yes, generating “abnormal” samples from your normal images can help, but only if the edits realistically match the true failure modes; otherwise you risk shortcut learning and label noise. Keep all synthetic generation strictly on the training split (never from val/test) and prefer a balanced, real-only validation set. If your end goal is simply abnormal vs normal, consider collapsing to a binary classifier; alternatively, retain 3 classes for training but merge the two “green_*” scores at inference. Because color is the key signal here, reduce color-jitter and strong auto-augmentation during training so the model does not learn color invariance:

```python
from ultralytics import YOLO

model = YOLO("yolo11n-cls.pt")
model.train(
    data="path/to/cls_data",  # folders per class
    imgsz=224,
    epochs=50,
    auto_augment=None,  # keep color fidelity
    hsv_h=0.0, hsv_s=0.0, hsv_v=0.0,  # disable color jitter
    erasing=0.0  # avoid occluding the key color cue
)
```

Also, add “hard negatives” (purple/blue near-green images labeled as normal) to tighten the boundary, and evaluate with per-class precision/recall or F1 rather than overall accuracy, which can be misleading on imbalanced sets; see why in our overview of [accuracy pitfalls on imbalanced datasets](https://www.ultralytics.com/glossary/accuracy) and tune augmentation using the [classification augmentation options](https://docs.ultralytics.com/guides/yolo-data-augmentation/#classification-specific-augmentations).

**Comment by liusse** [comment]
> Great question—yes, generating “abnormal” samples from your normal images can help, but only if the edits realistically match the true failure modes; otherwise you risk shortcut learning and label noise. Keep all synthetic generation strictly on the training split (never from val/test) and prefer a balanced, real-only validation set. If your end goal is simply abnormal vs normal, consider collapsing to a binary classifier; alternatively, retain 3 classes for training but merge the two “green_*” scores at inference. Because color is the key signal here, reduce color-jitter and strong auto-augmentation during training so the model does not learn color invariance:
> 
> from ultralytics import YOLO
> 
> model = YOLO("yolo11n-cls.pt")
> model.train(
>     data="path/to/cls_data",  # folders per class
>     imgsz=224,
>     epochs=50,
>     auto_augment=None,  # keep color fidelity
>     hsv_h=0.0, hsv_s=0.0, hsv_v=0.0,  # disable color jitter
>     erasing=0.0  # avoid occluding the key color cue
> )
> Also, add “hard negatives” (purple/blue near-green images labeled as normal) to tighten the boundary, and evaluate with per-class precision/recall or F1 rather than overall accuracy, which can be misleading on imbalanced sets; see why in our overview of [accuracy pitfalls on imbalanced datasets](https://www.ultralytics.com/glossary/accuracy) and tune augmentation using the [classification augmentation options](https://docs.ultralytics.com/guides/yolo-data-augmentation/#classification-specific-augmentations).


I am  glad to receive your reply!

I will try to disable these parameters during the training. Also, I have some questiones:

Firstly, can I directly generate abnormal data from normal class images, or will that affect the model's robustness? Or should I find a new batch of data to generate abnormal data?

Additionally, how can I evaluate using precision/recall or F1 score during training? Are there any parameters I can set for this?

Finally, using a confusion matrix, I can see the prediction results for each class in the validation/test set. However, I noticed that when testing the validation/test set with inference, the two results are inconsistent. For example, I found that all samples were predicted as the green_bar class. I noticed that the confusion matrix might have many misclassifications, but during inference, the misclassifications are relatively few. Is this because the data loading methods are different?

**Comment by liusse** [comment]
> Great question—yes, generating “abnormal” samples from your normal images can help, but only if the edits realistically match the true failure modes; otherwise you risk shortcut learning and label noise. Keep all synthetic generation strictly on the training split (never from val/test) and prefer a balanced, real-only validation set. If your end goal is simply abnormal vs normal, consider collapsing to a binary classifier; alternatively, retain 3 classes for training but merge the two “green_*” scores at inference. Because color is the key signal here, reduce color-jitter and strong auto-augmentation during training so the model does not learn color invariance:
> 
> from ultralytics import YOLO
> 
> model = YOLO("yolo11n-cls.pt")
> model.train(
>     data="path/to/cls_data",  # folders per class
>     imgsz=224,
>     epochs=50,
>     auto_augment=None,  # keep color fidelity
>     hsv_h=0.0, hsv_s=0.0, hsv_v=0.0,  # disable color jitter
>     erasing=0.0  # avoid occluding the key color cue
> )
> Also, add “hard negatives” (purple/blue near-green images labeled as normal) to tighten the boundary, and evaluate with per-class precision/recall or F1 rather than overall accuracy, which can be misleading on imbalanced sets; see why in our overview of [accuracy pitfalls on imbalanced datasets](https://www.ultralytics.com/glossary/accuracy) and tune augmentation using the [classification augmentation options](https://docs.ultralytics.com/guides/yolo-data-augmentation/#classification-specific-augmentations).

Hi！
I used a classification model to classify data, and when validating the data using a validation set, I want to obtain detailed information about misclassified data, such as their names. It seems that the existing parameters cannot achieve this functionality. Is there a simple and feasible way to implement this?

**Comment by liusse** [comment]
> Accuracy isn't useful when your dataset is imbalanced. You should have at least a balanced validation set.

I previously thought that the low training accuracy was due to an imbalanced dataset. Now, I have augmented the anomalous image data to match the scale of the normal images, but the accuracy is still problematic, and there are misclassifications. If possible, I can provide the dataset and would appreciate your help in identifying where the issue lies. Thank you very much; this is very important to me!

**Comment by glenn-jocher** [comment]
You can generate abnormal examples directly from your existing normal training images; you don’t need a separate batch, just be sure all synthetic data comes only from the training split (never from val/test) and that the edits realistically resemble your real anomalies so the model doesn’t learn artificial artifacts instead of true failures.  

For metrics: the current classification trainer reports accuracy by default; there is no flag to make training log precision/recall/F1 instead, so you’ll need to run `model.val()` periodically and compute those metrics yourself from the confusion matrix or saved predictions; the [classification task docs](https://docs.ultralytics.com/tasks/classify/) show how to run validation from CLI or Python.  

On the confusion-matrix vs inference mismatch: `val` and `predict` use the same core pipeline, so on the same images they should give identical labels as long as you fix the same weights file, image set, and `imgsz`; in practice big differences almost always come from using `last.pt` vs `best.pt`, a different folder of images, or different resizing/cropping, so I’d first double‑check those three match exactly.  

To get filenames of misclassified validation images, a simple Python loop over your val folders works, for example:  

```python
from ultralytics import YOLO
from pathlib import Path

model = YOLO("runs/classify/train/weights/best.pt")
val_root = Path("path/to/val")  # subfolders per class
names = model.names

mis = []
for class_dir in sorted(val_root.iterdir()):
    if not class_dir.is_dir():
        continue
    gt_name = class_dir.name
    for img_path in class_dir.glob("*.*"):
        r = model(str(img_path), imgsz=224, verbose=False)[0]
        pred_id = int(r.probs.top1)
        pred_name = names[pred_id]
        if pred_name != gt_name:
            mis.append((img_path, gt_name, pred_name))

print(f"Misclassified: {len(mis)}")
for p, gt, pr in mis[:20]:
    print(p, "gt:", gt, "pred:", pr)
```  

Since this issue is closed and marked stale, if accuracy is still poor after balancing the classes and tightening augmentation, please open a new issue with a small, shareable subset of your data plus the exact `yolo train`/`yolo val` commands so the team can investigate with a reproducible example.

**Comment by liusse** [comment]
> Since this issue is closed and marked stale, if accuracy is still poor after balancing the classes and tightening augmentation, please open a new issue with a small, shareable subset of your data plus the exact `yolo train`/`yolo val` commands so the team can investigate with a reproducible example.

"Thank you very much for your reply. I believe the classification is relatively reasonable now, but there may be some oversights. I will reduce the dataset and then open a new issue."

**Comment by glenn-jocher** [comment]
Sounds good—thanks for following up; a smaller, shareable subset plus the exact `yolo train`/`yolo val` commands (and ideally the `yolo checks` output) in a new issue will make it much easier for us to review and spot any remaining problems.
